runpod/tracer.py

```diff
@@ -0,0 +1,194 @@
+
+
+async def on_connection_reuseconn(
+    session, context, params: TraceConnectionReuseconnParams
+):
+    context.connect = asyncio.get_event_loop().time() - context.on_request_start
+
+
+async def on_request_chunk_sent(session, context, params: TraceRequestChunkSentParams):
+    if not hasattr(context, "payload_size_bytes"):
+        context.payload_size_bytes = 0
+    context.payload_size_bytes += len(params.chunk)
+
+
+async def on_response_chunk_received(
+    session, context, params: TraceResponseChunkReceivedParams
+):
+    if not hasattr(context, "response_size_bytes"):
+        context.response_size_bytes = 0
+    context.response_size_bytes += len(params.chunk)
+
+
+async def on_request_end(session, context, params: TraceRequestEndParams):
+    elapsed = asyncio.get_event_loop().time() - context.on_request_start
+    context.transfer = elapsed - context.connect
+    context.end_time = time()
+
+    # log to trace level
+    report_trace(context, params, elapsed)
+
+
+async def on_request_exception(session, context, params: TraceRequestExceptionParams):
+    context.exception = str(params.exception)
+    elapsed = asyncio.get_event_loop().time() - context.on_request_start
+    context.transfer = elapsed - context.connect
+    context.end_time = time()
+
+    # log to error level
+    report_trace(context, params, elapsed, log.error)
+
+
+def report_trace(context: types.SimpleNamespace, params, elapsed, logger=log.trace):
+    context.start_time = time_to_iso8601(context.start_time)
+    context.end_time = time_to_iso8601(context.end_time)
+    context.total = round(elapsed * 1000, 1)
+
+    if hasattr(context, "transfer") and context.transfer:
+        context.transfer = round(context.transfer * 1000, 1)
+
+    if hasattr(context, "connect") and context.connect:
+        context.connect = round(context.connect * 1000, 1)
+
+    if hasattr(context, "on_request_start"):
+        delattr(context, "on_request_start")
+
+    if hasattr(context, "trace_request_ctx"):
+        delattr(context, "trace_request_ctx")
+
+    if hasattr(params, "response") and params.response:
+        context.response_status = params.response.status
+
+    logger(json.dumps(vars(context)), context.request_id)
+
+
+def get_aiohttp_tracer() -> TraceConfig:
+    # https://docs.aiohttp.org/en/stable/tracing_reference.html
+    trace_config = TraceConfig()
+
+    trace_config.on_request_start.append(on_request_start)
+    trace_config.on_connection_create_end.append(on_connection_create_end)
+    trace_config.on_connection_reuseconn.append(on_connection_reuseconn)
+    trace_config.on_request_chunk_sent.append(on_request_chunk_sent)
+    trace_config.on_response_chunk_received.append(on_response_chunk_received)
+    trace_config.on_request_end.append(on_request_end)
+    trace_config.on_request_exception.append(on_request_exception)
+
+    return trace_config
+
+
+# Tracer for requests
+
+
+class TraceRequest:
```

Comment:
This could use a doc string explaining what this is for (and that it's a context manager).